### ISSUE TITLE
Best Hidden layer to compute sentence embedding similarity

### ISSUE THREAD

**Issue by simonefrancia** [facebookresearch/XLM#91]
Hi,
I would like to use XLM embeddings applied on couples of sentences in order to calculate a semantic similarity between them. 
I used the script you provide in order to get embeddings, but what is the best Transformer hidden layer to choose for my aim? 
After I chose one of them, what type of similarity would you recommend between sentences? For Laser we use normalized cosine similarity and it seemed to work pretty good.

Thanks

**Comment by simonefrancia** [comment]
Sorry @glample, could you give some quick advice about my question? Thanks

**Comment by glample** [comment]
Hi,

Sorry I don't really know, I never tried this. In any case you will need to finetune your model. You can decide to use the first hidden state, the average of the hidden states, or some element-wise max-pooling. You can try these 3 approches and see what works best.

**Comment by simonefrancia** [comment]
I tried the 3 approaches that you suggested without any finetuning. As similarity metric, I used this:
```
v1 = embed_xlm(text1)
v2 = embed_xlm(text2)
n1 = np.linalg.norm(v1)
n2 = np.linalg.norm(v2)
similarity = np.dot(v1, v2) / n1 / n2
```
that is the same I used for LASER.

I made some tests with some simple couples of sentences, that we used as benchmark, and it seems that all approaches don't work very well for semantic similarity. 
For example, for inputs:
EXAMPLE 1
```
"text1": "I was going to seaside",
"lang1": "en",
"text2": "cooking is one of my favorite activity",
"lang2": "en"
```
we get these similarities:
```
"similarity_hidden_layer_0": "0.764",
"similarity_sim_mean": "0.809",
"similarity_max_pool_element_wise": "0.887",
```

EXAMPLE 2
```
"text1": "tell me baby what is your story",
"lang1": "en",
"text2": "I need more hour with you",
"lang2": "en"
```

```
"similarity_hidden_layer_0": "0.825",
"similarity_sim_mean": "0.847",
"similarity_max_pool_element_wise": "0.825",

```
It's a bit strange because the semantic of the two sentences are different. 
How can I fine-tune, as you said in previous message?

**Comment by loretoparisi** [comment]
@glamoe just curious why the fine-tuning of the MLM + TLM model could improve on the STS task, I mean the embedding from the XNLI parallel corpus should be ok for this task?
My reference is the the LASER bi-LSTM model that has good results cross-lingual STS and it has been trained on the Tatoeba sentences corpus.

**Comment by glample** [comment]
@simonefrancia @loretoparisi The thing is that XLM is not really designed to have a single embedding at any point, as the task was not to create a single vector to represent a full sentence like in LASER. In LASER you may not need fine-tuning because the model is trained to have single vector sentence representations. Here a sentence is represented by a list of hidden states. You can do some average pooling or max pooling as a trick to transform the list of vectors to a single vector, but this will never be as good as nothing guarantees that the averaged vectors is a good representation. This is why I think it's necessary to do some fine-tuning on the downstream task if you want to have a single vector representation.

**Comment by loretoparisi** [comment]
@glample thank you very much. So we can say that LASER fits better for STS tasks because of the learned high dimension embedding that are different than XLM (single vector against hidden states). We were also thinking to use XLM for zero shot classification, a task in which we were running LASER with sub-optimal results (https://github.com/facebookresearch/LASER/tree/master/tasks/mldoc). Do you think that XLM can improve the baseline on MLDoc dataset in this case?
Thank you.

**Comment by glample** [comment]
If you train on the full sentence representations, i.e. if you have a classifier that takes as input not a single vector but a sequence of vectors, then XLM should work at least as well. I'm pretty sure it should work better because you have more information before you do the classification, since the sentence is not compressed in any way.

If you are interested in doing interpolation between sentences, then LASER is probably better since you will really want to have a single vector representation. But for classification, you have no reason to compress your sentence in just one vector.

I'm not sure about MLDoc sorry, I don't know this dataset at all.
